### PR TITLE
Enable multi-producer single-consumer mode in linuxmodule FromDevice

### PR DESCRIPTION
**Pull request by MappaM**
If using multiqueue with irqbalance (or manual pinning of each queue interrupt to multiple cores to share load) the enqueueing function will be called on the interrupt's core.

The producer use a CAS instruction to grab the head in the ring and advance it. As the consumer could read the ring[head] before the producer writes to it, the 0 value is used to specify a "writing pending" and the consumer spins until it reads a != 0 value.

For instance, it leads to a bsod with an ixgbe NIC in an out of the box debian configuration, where ixgbe enables MQ/RSS by default and irqbalance kicks in.

This is a simple bug-fix, but a proper support of multiqueue would be a better solution... I don't think we need kernel patching (which probably prevented merge from routebricks and vrouter?) anymore as from kernel 3.2 we have the skb->queue_mapping available. Speaking of it, having one internal queue in the fromdevice for each different "queue_mapping" seen, and having the run_task checking all queue would probably be quicker than the solution proposed in this patch using a costly CAS...

### CODE REVIEW

**Comment by MappaM** [comment]
I've rebased on the last branch, it's more clear now thanks to the modifications following #154

File: elements/linuxmodule/fromdevice.cc [hunk 1]
```diff
@@ -472,14 +471,19 @@ FromDevice::run_task(Task *)
...
+        emission_report(h);
 #endif
-	set_head(next_i(h));
-	output(0).push(p);
-	npq++;
-	_count++;
+        set_head(next_i(h));
+        output(0).push(p);
+        npq++;
+        _count++;
```
Comment (kohler) [inline elements/linuxmodule/fromdevice.cc]: It isn't super great when a commit includes both whitespace changes and technical changes.

File: include/click/standard/storage.hh [hunk 2]
```diff
@@ -5,10 +5,13 @@
 CLICK_DECLS
 class Packet;
 
+#include <click/atomic.hh>
```
Comment (kohler) [inline include/click/standard/storage.hh]: `#include`s don't belong in CLICK_DECLS, this should be above line 5

File: include/click/standard/storage.hh [hunk 3]
```diff
@@ -21,6 +24,17 @@ class Storage { public:
...
     index_type tail() const		{ return _tail; }
 
+    index_type take_tail() {
+        index_type shadow_tail;
+        index_type next_shadow_tail;
+        do {
+            shadow_tail = _tail;
+            next_shadow_tail = next_i(shadow_tail);
+            if (next_shadow_tail == _head) return index_error;
+        } while (shadow_tail != reinterpret_cast<atomic_uint32_t*>(const_cast<index_type*>(&_tail))->compare_swap(shadow_tail,next_shadow_tail));
```
Comment (kohler) [inline include/click/standard/storage.hh]: Prefer this line:

```
} while (shadow_tail != atomic_uint32_t::compare_swap(_tail, shadow_tail, next_shadow_tail));
```

i.e. use the static version of atomic_uint32_t::compare_swap.

File: include/click/standard/storage.hh [hunk 4]
```diff
@@ -5,10 +5,13 @@
 CLICK_DECLS
 class Packet;
 
+#include <click/atomic.hh>
+
 class Storage { public:
 
     typedef uint32_t index_type;
     typedef int32_t signed_index_type;
+    #define index_error UINT_MAX
```
Comment (kohler) [inline include/click/standard/storage.hh]: NO `#define`s!!!!
```
static const index_type index_error = (index_type) -1;
```

File: elements/linuxmodule/fromdevice.hh [hunk 5]
```diff
@@ -131,6 +131,7 @@ Resets the count and drops handlers.
 
 #include "elements/linuxmodule/anydevice.hh"
 #include <click/standard/storage.hh>
+#include <click/sync.hh>
```
Comment (kohler) [inline elements/linuxmodule/fromdevice.hh]: Is this needed?

File: elements/linuxmodule/fromdevice.cc [hunk 6]
```diff
@@ -389,14 +392,12 @@ FromDevice::got_skb(struct sk_buff *skb)
...
-    if (nt != h) { /* ours */
-	skb = skb_share_check(skb, GFP_ATOMIC);
-	if (!skb)
-	    return 1;
-	assert(skb_shared(skb) == 0);
+    skb = skb_share_check(skb, GFP_ATOMIC);
+    if (!skb)
+        return 1;
+    assert(skb_shared(skb) == 0);
+    unsigned my_tail = take_tail();
```
Comment (kohler) [inline elements/linuxmodule/fromdevice.cc]:
```
Storage::index_type t = atomic_reserve_tail();
```

or something like it.

Comment (kohler) [inline elements/linuxmodule/fromdevice.cc]: Also please don't say `my_tail`; `t` is just as good and requires fewer changes.

**Comment by kohler** [comment]
OK, I really like this, but there are enough style issues that I'm going to bite off a piece (the storage.hh piece) and commit it on my own with credit to you.

After that's in, please rebase?

**Comment by kohler** [comment]
See daec752

**Pull request closed without merging**